XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,309 @@
+| Title  | AsyncWrap API |
+|--------|---------------|
+| Author | @trevnorris   |
+| Status | DRAFT         |
+| Date   | 2016-04-05    |
+
+## Description
+
+Since its initial introduction with the `AsyncListener` API, `AsyncWrap` has
+slowly evolved to ensure a generalized API that would serve as a solid base for
+module authors who wished to add events to the life cycle of the event loop.
+The API presented below aims to accomplish this.
+
+## API
+
+### Overview
+
+```js
+// Standard way of requiring a module. Snake case follows core module
+// convention.
+const async_wrap = require('async_wrap');
+
+// List of asynchronous providers (e.g. CRYPTO or TCPWRAP). Can be used for
+// trace filtering of events.
+const providers = async_wrap.Providers;
+
+// init() is called during object construction. The object handle being
+// constructed can be accessed as "this". Because init() is called during
+// construction it is not safe to blindly inspect the constructed object at the
+// moment. Though we are currently investigating how to make this safe.
+function init(id, type, parentId) { }
+
+// pre() is called just before the handle's callback is called. It can be
+// called 0-N times for handles (e.g. TCPWRAP), and should be called exactly 1
```

Comment:
It's not an implementation detail. It's a concrete detail in how node works. handles do work via requests. e.g. when a TCP server handle receives a new connection, it receives a new connection _request_. which is then converted into a handle that can receive requests like new data packet. this has an important distinction when wanting to track metrics and resources.

and just because an object isn't an instance of `ReqWrap` doesn't mean it doesn't fit in this model. hypothetically all async activity could be made a subclass of `HandleWrap` or `ReqWrap`, but we don't for performance reasons.

> There is no "object" being constructed during init().

`TickObject()` is constructed for every call to `nextTick()`. Promise has an internal construction mechanism. Promises construct a new object. [The ECMA spec](http://www.ecma-international.org/ecma-262/6.0/index.html#sec-promise-executor) shows that each new promise has several internal fields set for each newly constructed instance.

But technically they would each be a request. They only trigger once (Promises will fire on multiple provided resolve/reject callbacks, but only once) and they themselves will never hold references to another request.